dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java

```diff
@@ -637,60 +568,48 @@ private String getFromSubQuery( TrackedEntityInstanceQueryParams params, boolean
      */
     private String getFromSubQuerySelect( TrackedEntityInstanceQueryParams params )
     {
-        StringBuilder subqueryBuilder = new StringBuilder();
-
-        subqueryBuilder
-            .append( "SELECT " )
-            .append( "TEI.trackedentityinstanceid, " )
-            .append( "TEI.uid, " )
-            .append( "TEI.created, " )
-            .append( "TEI.lastupdated, " )
-            .append( "TEI.inactive, " )
-            .append( "TEI.trackedentitytypeid, " )
-            .append( "TEI.potentialduplicate, " )
-            .append( "TEI.deleted, " )
-            .append( "OU.uid as ou, " )
-            .append( "OU.name as ouname " );
-
-        subqueryBuilder.append( getSubqueryFromClauseOrderFields( params, subqueryBuilder.toString() ) );
-
-        return subqueryBuilder.toString();
-    }
-
-    /**
-     * OrderColumn Generates a string containing the columns for projections
-     * based on the params we are ordering by.
-     *
-     * @param params
-     * @return a string containing the columns to be used in the subquery SQL
-     *         projection, or empty string if no attributes nor order params are
-     *         used for ordering.
-     */
-    private String getSubqueryFromClauseOrderFields( TrackedEntityInstanceQueryParams params, String query )
-    {
-        StringBuilder orderAttributes = new StringBuilder();
+        LinkedHashSet<String> columns = new LinkedHashSet<>(
+            List.of( "TEI.trackedentityinstanceid", "TEI.uid", "TEI.created", "TEI.lastupdated",
+                "TEI.inactive", "TEI.trackedentitytypeid", "TEI.potentialduplicate", "TEI.deleted", "OU.uid as ou",
+                "OU.name as ouname " ) );
 
         for ( OrderParam orderParam : params.getOrders() )
         {
-            if ( isStaticColumn( orderParam.getField() ) )
+            Optional<TrackedEntityInstanceQueryParams.OrderColumn> orderColumn = getColumn( orderParam.getField() );
+
+            if ( orderColumn.isPresent() )
             {
-                orderAttributes.append( addStaticColumn( orderParam.getField(), query ) );
+                columns.add( orderColumn.get().getTableAlias() + "." + orderColumn.get().getColumn() );
```

Comment:
Maybe it would make sense to have a method in the OrderColumnd enum that build this, we could reuse it while constructing the order by sql too. Something like:

```
public static String getColumnStatement() {
    return getTableAlias() + "." + getColumn()
} 
```